api/v1_projects.inc

```diff
@@ -752,3 +753,116 @@ function _get_enabled_filter($query_params)
 {
     return get_flag_value($query_params, "enabled");
 }
+
+//---------------------------------------------------------------------------
+// Proofreading functions
+
+// checkout a page
+function api_v1_project_checkout($method, array $data, array $query_params)
+{
+    try {
+        $proof_project = create_proof_project($data, $query_params);
+        return $proof_project->checkout();
+    } catch (ProjectException $exception) {
+        throw new NotFoundError($exception->getMessage(), $exception->getCode());
+    } catch (UserAccessException $exception) {
+        throw new ForbiddenError($exception->getMessage(), $exception->getCode());
+    }
+}
+
+function api_v1_project_validatetext($method, $data, array $query_params)
+{
+    $text_checker = new TextChecker();
+    return $text_checker->analyse(receive_project_text_from_request_body('text'), $data[":projectid"]);
+}
+
+function api_v1_project_page(string $method, array $data, array $query_params)
+{
+    global $pguser;
+
+    try {
+        $proof_project = create_proof_project($data, $query_params);
+        $project_page = $data[":pagename"];
+        $page_state = $query_params['pagestate'] ?? null;
+
+        if (null === $page_state) {
+            throw new InvalidValue("No page state found in request.");
+        }
+        if (!in_array($page_state, Rounds::get_page_states())) {
+            throw new InvalidValue(sprintf("%s is not a valid page state", $page_state));
+        }
+        if ($page_state != $project_page->page_state) {
+            $err = sprintf(
+                _('Warning: Page  "%1$s" is no longer in state "%2$s"; it is now in state "%3$s".'),
+                $project_page->page_name,
+                $page_state,
+                $project_page->page_state
+            );
+            throw new ProjectPageInconsistentStateException($err);
+        }
+
+        $proof_project_page = new ProofProjectPage($proof_project, $project_page);
+        if ($method == "GET") {
```

Comment:
From a usability perspective, the GET shouldn't need to pass any any of the state information to get the page text. And it should probably return the page state in the return -- probably the `page_text_state` object, actually.

Reply:
This function can be used for the 'revert to last save' in the enhanced PI. If the PM changes the project to unavailable while a user has a page open in his PI then at present if he does a 'revert to last save'  he gets a warning that the project has changed state and is unable to proceed. Do we need a different function to get the text etc. for non-proofreading purposes?

Reply:
I see what you mean about the 'GET'. Let me think about that.

Reply:
In fact we don't need that GET function at all. It was intended to be used for the 'revert to last save' in the enhanced PI, but you can get the same effect in the standard PI by 'stop proofreading' and resuming the page so we can use the resume_page PUT function instead.
That leaves the question of the PUT functions. They can all change the database in some way so I think we need to check that the project is available as you said.

Reply:
No no, we need the GET. Because a GET is the only way we have to get the state of a page that we need to do anything with it. Without the GET we are blind to how to interact with the page.

Reply:
It seems like you are envisaging a different use  for the GET function than I had anticipated. I may be going about it the wrong way but the following is the workflow I am considering. (using the new funcion names)

Checkout a page. Returns all page data or an exception if it can't be done.

Save. (formerly as_in_progress) returns the text (could be more normalised than what was submitted) and page state which could have changed. Throws exception if it contains invalid characters.

Revert. Saves submitted text and returns original. (This supports the coresponding function in enhanced PI. The standard PI can't do this)

Checkin. Saves page as done.

Abandon. Return page to the round.

Resume. Returns all page data. If the page was in the saved_as_done state, changes it to 'temp' and makes appropriate changes to the database.

Reply:
That workflow works great if it's done all in one "session". But consider the case where the API client:
1. checks out page (gets 001)
2. saves page 001
3. checks out another page (gets 002)
4. saves page 002
5. checks in page 002

If I'm the client in order to do something with 001 I have to keep track of what the last state of the page is (and the text itself). But if I have a GET function I could just get the pagetext and  state and then save it or check it in or whatever.

Put another way, if we have the GET function for a page, then we're just one more endpoint away (one that returns all of the pages the user has worked on in a project in the current round) from the client not needing to retain any state beyond a page it's currently working with.
1. client GETs list of all pages the user has worked on in the project in the current round (new endpoint, we should not add it to this PR)
2. client can GET a page they want to do something with (checkout, revert, save, whatever)
3. client now has the data it needs to do does something with that page
4. client does some action with that page (checkout, revert, save, etc)
5. lather, rinse, repeat

Reply:
I'd only considered it as if interfacing from the existing project page where you either checkout a page or access a page you have worked on before. If you want to go further back then, as you say, more is required. I'll consider this.

Reply:
The new endpoint could return a list of the pages the user has worked on together with their states and timestamps. Similar to the code used to make the lists of pages the user has worked on in the project page. The client could make similar lists and the user could select a page and do what he wanted using the existing endpoints.

For this new endpoint the question arises of how the client would understand the page state? Should we also supply a more concise string than the one used in the database, perhaps just "done", "temp" or "out" (which are the only relevant ones here and the endpoint would have explicitly selected)

Reply:
I think I see the disconnect. We aren't designing the API to replicate the UX we already have. We're designing an API that can be used as a data layer primarily, but not exclusively for, a new proofreading interface. The API should "stand on its own" in a sense.

I'm not at all prescribing what we should present to the user, etc -- more that as an API consumer I want to be able to use these new endpoints stand-alone as a data transport. We should 100% keep the needs of a future API as our primary design focus, but we should also make sure the API endpoints are useful outside of that.

For "list all pages that I worked on in this project" endpoint, it should return the same set of data that the client then needs to work with the PUT endpoint -- so the exact same state name.

Reply:
But let's not add that new endpoint to this PR. Let's focus on wrapping up the endpoints you have.

Reply:
I'll resolve the conflicts and push it again.

Reply:
Any progress on adding the GET endpoint back? I'd been waiting on that to re-review this.

Reply:
You are correct, it did not do what I wanted but we still need it. It needs to not require passing in any state (project or page) and it should return a `page_text_state` object so that it includes the page text and the current page state. We need a GET to get the current state of a page so I can do something with it. I can't call resume without the page state.

Separately, and not in this PR, we need an endpoint that will return all pages for a specific user that the API could use to present a user with pages they have done in the past and can work on again.

Reply:
I can do that.
I've been maintaining a basic PI based on this branch to show how the functions can be used. I've pushed it up now.